enhancements/testing/ephemeral-agent-installer-e2e-tests.md

```diff
@@ -0,0 +1,275 @@
+This is mostly because there isn't sufficient information about the actual failure when 
+it happens and there are tons of red herring errors in the logs. There has been significant
+efforts in trying to fix these issues in the past but they almost never get prioritized
+over new feature development. With every new feature added the complexity grows and so does
+the knowledge gap of the e2e tests and the average developer on the team. With this new 
+project I would like to avoid this outcome as much as possible.
+
+### Goals
+
+This proposal will focus using this new e2e framework in the new ephemeral agent based installer 
+image project only. 
+
+### Non-Goals
+
+* Re-factoring dev-scripts and assisted-test-infra for reasons not related to the new 
+ephemeral agent based installer image project.
+* Getting buy in from other installer teams (Assisted Installer, IPI) to use this new 
+framework in their e2e tests for the upcoming release.
+
+This is not to say that we shouldn't talk to other installer teams about this and 
+collaborate with them. It is just not a goal to have them use the same e2e framework as the 
+new project.
+
+## Proposal
+
+Create an e2e testing framework for the ephemeral agent based installer image.
+
+### User Stories
+
+#### Run Assisted Installer Outside k8s
+
+As a developer, I need to run Assisted Installer without installing a k8s platform first 
+so that I can test the installation of a users first cluster, cluster 0.
+
+This is the overall purpose of this ephemeral agent based installer image project. 
+Unfortunately, neither of the existing frameworks actually can actually assist with this. 
+
+* The Assisted Installer e2e tests that run in dev-scripts first use the IPI installer 
+to install OCP and then the infrastructure-operator on this OCP cluster making it a hub 
+cluster.
+
+* The Assisted Installer e2e tests that run in assisted-test-infra installs minikube first 
+and then runs Assisted Installer on top of it.
+
+Which means in both existing e2e frameworks we first setup some form of k8s first. Given
+the goal of the ephemeral agent based installer project to install a users first cluster, 
+cluster 0, this is highly undesirable.
+
+While there is a flow to run Assisted Installer in podman this is a manually supported 
+flow. The support of it in the e2e frameworks is mostly abandoned by the Assisted Installer 
+team in favor of infrastructure-operator and the hosted SaaS solution.
+
+#### Setup Environment for Install Without Installing OCP or k8s
+
+As a developer, I need to set up VMs so that I can simulate the installation of a users 
+first cluster using the ephemeral based agent installer image.
+
+Both existing frameworks today set up VMs using libvirt today to test a users OCP install.
+However, they also both immediately use those VMs to start the installation process. There
+isn't an easy way to stop the current frameworks from using IPI and Assisted Installer
+after the VM creation.
```

Comment:
This is not accurate, in dev-scripts you can simply run [01_install_requirements.sh](https://github.com/openshift-metal3/dev-scripts/blob/master/01_install_requirements.sh) then [02_configure_host.sh](https://github.com/openshift-metal3/dev-scripts/blob/master/02_configure_host.sh)

From that point you can do whatever you want with the VMs, for example testing some new installer workflow (which could be added as a new script)

Another option if you really don't want to use the same scripts is to just consume the upstream [metal3-dev-env ansible roles](https://github.com/metal3-io/metal3-dev-env/tree/master/vm-setup) which we use inside dev-scripts - this might be preferable if you're aiming for a pure-ansible solution.

Reply:
It would seem unfortunate indeed if we were to end up rewriting the vm-setup roles from metal3-dev-env.

This might even be an opportunity to start converting more of the first couple of scripts in dev-scripts to ansible, which would benefit both projects.

Reply:
I agree that the roles are decoupled enough that they could be used independently of the deployment steps. There are other questions about their fitness that we could answer here, though.

Do they already do something close enough to what is needed that they could be used without other significant refactoring?

Is there any concern that reusing metal3-dev-env would tie this project to the expectations or assumptions of metal3 upstream?

Reply:
This section is still not accurate. We've established that at least dev-scripts has stages and that those stages can be run independently so that a cluster isn't deployed via IPI.

Reply:
I have removed this paragraph.